### PR TITLE
Use shape placeholder in example

### PR DESCRIPTION
**Pull request by okuta**
Update examples and documentation to use shape placeholder.

### CODE REVIEW

**Comment by delta2323** [comment]
> This PR still includes bug.

Then, could you add [WIP] to the title?

**Comment by okuta** [comment]
I fixed bugs. Please review this PR.

**Comment by gwtnb** [comment]
`train_mnist_data_parallel.py` failed.

**Comment by zori** [comment]
I created a PR that fixes `train_mnist_data_parallel.py`. I have tested the code with and without parameter shape placeholder. My proposed fix is https://github.com/pfnet/chainer/pull/1576

**Comment by okuta** [comment]
Thank you @gwtnb and @zori .
I merged and fixed.

File: chainer/training/updater.py [hunk 1]
```diff
@@ -307,6 +308,10 @@ def update_core(self):
                 in_vars = variable.Variable(in_arrays)
                 losses.append(loss_func(in_vars))
 
+        # For _uninitialized_params
```
Comment (zori) [inline chainer/training/updater.py]: I think this call to cleargrads(), just before the backward pass, is enough for all parameters. There is no need to clear them here: https://github.com/pfnet/chainer/blob/0ceb2b20d1cd7b16b90ceb8e78d55774e77777cb/chainer/training/updater.py#L292-L293 before the forward pass.

**Comment by zori** [comment]
I updated the tutorial https://github.com/pfnet/chainer/pull/1584

File: chainer/training/updater.py (hunk 1 above)
Comment (okuta) [inline chainer/training/updater.py]: `add_param` fill `grad` with `nan`.
https://github.com/pfnet/chainer/blob/master/chainer/link.py#L141

Comment (disktnk) [inline chainer/training/updater.py]: Why `clear_grads` is set before call loss_func?
https://github.com/pfnet/chainer/pull/1551/files#diff-32a6fe336029b64713fe8980fe17f085R293

**Comment by bkvogel** [comment]
LGTM!

**Pull request merged**